Client/Frontend/Browser/Playlist/BrowserViewController+Playlist.swift

```diff
@@ -7,81 +7,145 @@ import Foundation
 import Data
 import Shared
 import BraveShared
+import BraveUI
 
 private let log = Logger.browserLogger
 
 extension BrowserViewController: PlaylistHelperDelegate {
-    func showPlaylistAlert(_ alertController: UIAlertController) {
-        self.present(alertController, animated: true)
-    }
     
-    func showPlaylistToast(info: PlaylistInfo, itemState: PlaylistItemAddedState) {
-        guard Preferences.Playlist.showToastForAdd.value,
-              let selectedTab = tabManager.selectedTab,
-              selectedTab.url?.isPlaylistSupportedSiteURL == true else {
-            return
+    func updatePlaylistURLBar(tab: Tab?, state: PlaylistItemAddedState, item: PlaylistInfo?) {
+        openInPlayListActivity(info: state == .existingItem ? item : nil)
+        addToPlayListActivity(info: state == .newItem ? item : nil, itemDetected: state == .newItem)
+        
+        switch state {
+        case .none:
+            topToolbar.menuButton.removeBadge(.playlist, animated: true)
+            toolbar?.menuButton.removeBadge(.playlist, animated: true)
+        case .newItem, .existingItem:
+            topToolbar.menuButton.addBadge(.playlist, animated: true)
+            toolbar?.menuButton.addBadge(.playlist, animated: true)
         }
         
-        if let toast = playlistToast {
-            toast.item = info
+        if let selectedTab = tabManager.selectedTab {
+            selectedTab.playlistItemState = state
+            selectedTab.playlistItem = item
+            
+            let shouldShowPlaylistURLBarButton = tab?.url?.isPlaylistSupportedSiteURL ?? false
+            let playlistButton = topToolbar.locationView.playlistButton
+            switch state {
+            case .none:
+                playlistButton.buttonState = .none
+            case .newItem:
+                playlistButton.buttonState = shouldShowPlaylistURLBarButton ? .addToPlaylist : .none
+            case .existingItem:
+                playlistButton.buttonState = shouldShowPlaylistURLBarButton ? .addedToPlaylist : .none
+            }
+        } else {
+            topToolbar.locationView.playlistButton.buttonState = .none
+            topToolbar.menuButton.removeBadge(.playlist, animated: true)
+            toolbar?.menuButton.removeBadge(.playlist, animated: true)
+        }
+    }
+    
+    func showPlaylistPopover(tab: Tab?, state: PlaylistPopoverState) {
+        guard Preferences.Playlist.showToastForAdd.value,
+              let selectedTab = tabManager.selectedTab else {
```

Comment:
we pass `tab` in this method, is it different from `selectedTab`?
I wonder if we could skip using `selectedTab` in this method

Reply:
Yeah it can be different. This function is called from a JS callback which doesn't know anything about the tab and whether it is selected or not. So I have guarded against that in the latest PR so it doesn't show the popover on the wrong tab.